include/revng/ADT/GenericGraph.h

```diff
@@ -10,8 +10,73 @@
 #include "llvm/ADT/STLExtras.h"
 #include "llvm/ADT/SmallVector.h"
 
+#include "revng/ADT/STLExtras.h"
 #include "revng/Support/Debug.h"
 
+/// GenericGraph is our implementation of a universal graph architecture.
+/// First and foremost, it's tailored for the revng codebase, but maybe you
+/// can find it useful as well, that's why it's released under the MIT License.
+///
+/// To use the Graph, you need to make you choice of the node type. We're
+/// currently supporting three architectures:
+///
+///   - ForwardNode - the most trivial one. It uses the least memory, but each
+///                   node only stores links with its successors. That makes
+///                   the inverse graph iteration expensive.
+///
+///   - BidirectionalNode - A simple double-linked node architecture.
+///                         Each node remembers links with both its successors
+///                         and predecessors. Take note, that for the sake of
+///                         implementation simplicity this node stores COPIES
+///                         of the labels. It's only suitable to be used with
+///                         cheap to copy labels which are never actually
+///                         mutated. There are plans on making them explicitly
+///                         immutable (TODO). So you can consider node mutation
+///                         a deprecated behaviour.
+///
+///    - MutableEdgeNode - The most complex node type we're supporting.
+///                        It is similar to BidirectionalNode except it stores
+///                        the label on the heap. It's slower and more memory
+///                        consuming but it allows for safe edge label
+///                        modification as well as controls that nodes and
+///                        edges are safely removed so no leftover parts are
+///                        ever present.
```

Comment:
> The most complex node type we're supporting

A bit scary :)

Otherwise :+1:.

Reply:
Well, compared to other two :)

"The most complex" doesn't mean much when the other two alternatives are cute little plush kittens that don't even allocate :)

Reply:
Done :)